MarkBind/markbind
# Add informative error for cyclic dependency

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Other, please explain: Not exactly a bug fix / new feature, but an enhancement to current error message. Should I tag it under Enhancement?

Resolves #594.

**What is the rationale for this request?**
Currently, authors might unintentionally create cyclic references in their `<include />` tag. For example, two files 
a.md
```
<include src="b.md" />
```
and b.md
```
<include src="a.md" />
```
would cause MarkBind to error with an uninformative RangeError: Maximum call stack size exceeded.

**What changes did you make? (Give an overview)**
I created a custom error object (that extends Javascript's Error), which keeps track of the last 5 files processed before an error occurred. This helps the user identify the cyclic dependency loop.
When a cyclic dependency occurs, users would see
```
error: Cyclic reference detected.
Error while preprocessing '/path/to/file/a.md'
Last 5 files processed: 
        /path/to/file/a.md
        /path/to/file/b.md
        /path/to/file/a.md
        /path/to/file/b.md
        /path/to/file/a.md
```
**Is there anything you'd like reviewers to focus on?**
I'm not sure if this is the best way to implement cyclic dependency checks. Please let me know if you can think of a better way to keep track and print visited files, or if creating the custom error class is not appropriate.
**[NEW]** I've implemented our custom call stack to keep track of the files. This enables me to throw an error when its above 200 items, and additionally keep track of the last 5 files. As a future extension, perhaps we can extend this into a full-blown file manager class.
**[NEW]** thanks @nicholaschuayunzhi : I realise there's no issue when including dynamic cyclic reference during the preprocess stage. Dynamic references cause the error only when Page.js tries to `collectHeadingsAndKeywordsInContent`. Thanks for helping me spot the gap; I've also added try/catch on the error there, although the error message is not as informative because there's no longer a recursion tree.

**Testing instructions:**
Ensure that you have fetched my PR, then create two files: `a.md` and `b.md` with contents as above.
